### ISSUE TITLE
DownloadManager cannot specify custom CacheKeyFactory

### ISSUE THREAD

**Issue by saiday** [google/ExoPlayer#5013]
If I use `DownloadManager` for downloading, there is no way I can assign custom `CacheKeyFactory`.

Because `DownloadManager` heavily relies on `DownloaderConstructorHelper`.
`DownloaderConstructorHelper`'s `buildCacheDataSource()` which is `DownloaderManager` generated `Dwonloader` used to build as `Downloader` datasource, has no way to interrupt with custom `CacheKeyFactory`.

```java
  public CacheDataSource buildCacheDataSource(boolean offline) {
    DataSource cacheReadDataSource = cacheReadDataSourceFactory != null
        ? cacheReadDataSourceFactory.createDataSource() : new FileDataSource();
    if (offline) {
      return new CacheDataSource(cache, DummyDataSource.INSTANCE,
          cacheReadDataSource, null, CacheDataSource.FLAG_BLOCK_ON_CACHE, null);
    } else {
      DataSink cacheWriteDataSink = cacheWriteDataSinkFactory != null
          ? cacheWriteDataSinkFactory.createDataSink()
          : new CacheDataSink(cache, CacheDataSource.DEFAULT_MAX_CACHE_FILE_SIZE);
      DataSource upstream = upstreamDataSourceFactory.createDataSource();
      upstream = priorityTaskManager == null ? upstream
          : new PriorityDataSource(upstream, priorityTaskManager, C.PRIORITY_DOWNLOAD);
      return new CacheDataSource(cache, upstream, cacheReadDataSource,
          cacheWriteDataSink, CacheDataSource.FLAG_BLOCK_ON_CACHE, null);
    }
  }
```
https://github.com/google/ExoPlayer/blob/release-v2/library/core/src/main/java/com/google/android/exoplayer2/offline/DownloaderConstructorHelper.java#L90-L106

---

I'm not sure if I'm asking A/B question, my main goal is handling the identically same file but multiple HTTP URIs that only URI's HOST modified.
Our backend will give multiple URI that the file is identically same but from different CDN provider.

I was trying to store URI's path as cache key

**Comment by erdemguven** [comment]
This has been fixed but not released yet unfortunately:
https://github.com/google/ExoPlayer/blob/dev-v2/library/core/src/main/java/com/google/android/exoplayer2/offline/DownloaderConstructorHelper.java#L100

CacheKeyFactory is the right solution for your problem.

**Comment by saiday** [comment]
Thanks for pointing out the commit. Waiting for the release!

**Comment by saiday** [comment]
ExoPlayer 2.9.1 had been released, not including this patch unfortunately.
@erdemguven Do you have any chance to know when this patch will be released? I'm planning an app update depend on this. 

Thank you.

**Comment by erdemguven** [comment]
@ojw28 Is there a blocker for this change?

**Comment by ojw28** [comment]
It wasn't included in 2.9.1 only because it looked like quite a big change for a minor release (and not sufficiently important). I can take a look at putting it in 2.9.2, but it'll probably be in 2.10.0 instead. There is no timeline for this release currently.

**Comment by saiday** [comment]
@ojw28 thank you for clarification
I totally understand your release decision.

Here are my two cents, this patch is essential for custom cache key if using `DownloadManager`. I haven't found an alternative way to achieve the purpose.
Since `DownloadManager` is your newly recommended way to download streams ([Downloading Streams](https://medium.com/google-exoplayer/downloading-streams-6d259eec7f95)), I might not want to fall back to `Downloader` if this patch would be released in recent times.

Please kindly let us know If this patch would be scheduled to 2.10.0 after taking a look at 2.9.2 possibility.
 
Thank you for your great efforts.

**Comment by ojw28** [comment]
The change isn't a complete solution in any case, since it doesn't use the custom keys when removing the download, or when scanning the cache to see what's been downloaded.

**Comment by saiday** [comment]
@erdemguven Is there anything I can help?

**Comment by ojw28** [comment]
The change above makes the solution complete. There is no possibility of this functionality being included in a 2.9.x release. It will be in 2.10.0, for which there is not currently an ETA.

**Comment by saiday** [comment]
@ojw28 I appreciated your time and effort, will waiting for the release.
